pkg/storage/stores/shipper/compactor/index_set.go

```diff
@@ -0,0 +1,337 @@
+}
+
+// newUserIndex intializes a new index set for user index. Other than creating instance of indexSet, it also compacts down the source index.
+func newUserIndex(ctx context.Context, tableName, userID string, baseUserIndexSet storage.IndexSet, workingDir string, logger log.Logger) (*indexSet, error) {
+	if !baseUserIndexSet.IsUserIndexSet() {
+		return nil, fmt.Errorf("base index set is not for user index")
+	}
+
+	if err := util.EnsureDirectory(workingDir); err != nil {
+		return nil, err
+	}
+
+	ui := &indexSet{
+		ctx:          ctx,
+		tableName:    tableName,
+		userID:       userID,
+		workingDir:   workingDir,
+		baseIndexSet: baseUserIndexSet,
+		logger:       log.With(logger, "user-id", userID),
+		ready:        make(chan struct{}),
+	}
+
+	ui.initUserIndexSet(workingDir)
+	return ui, nil
+}
+
+// initUserIndexSet downloads the source index files and compacts them down to a single index file.
+func (is *indexSet) initUserIndexSet(workingDir string) {
+	defer close(is.ready)
+	ctx, cancelFunc := context.WithTimeout(is.ctx, userIndexReadinessTimeout)
+	defer cancelFunc()
+
+	is.sourceObjects, is.err = is.baseIndexSet.ListFiles(is.ctx, is.tableName, is.userID)
+	if is.err != nil {
+		return
+	}
+
+	compactedDBName := filepath.Join(workingDir, fmt.Sprint(time.Now().Unix()))
+	seedFileIdx := findSeedFileIdx(is.sourceObjects)
+
+	if len(is.sourceObjects) > 0 {
+		// we would only have compacted files in user index folder, so it is not expected to have -1 for seedFileIdx but
+		// let's still check it as a safety mechanism to avoid panics.
+		if seedFileIdx != -1 {
+			seedFileIdx = 0
+		}
+		compactedDBName = filepath.Join(workingDir, is.sourceObjects[seedFileIdx].Name)
+		is.err = shipper_util.DownloadFileFromStorage(func() (io.ReadCloser, error) {
+			return is.baseIndexSet.GetFile(ctx, is.tableName, is.userID, is.sourceObjects[seedFileIdx].Name)
+		}, shipper_util.IsCompressedFile(is.sourceObjects[seedFileIdx].Name), compactedDBName, false, is.logger)
+		if is.err != nil {
+			return
+		}
+
+		// if we have more than 1 file, we would compact it down so upload the compacted DB and remove the source objects
+		if len(is.sourceObjects) > 1 {
+			is.uploadCompactedDB = true
+			is.removeSourceObjects = true
+		}
+	}
+
+	is.compactedDB, is.err = openBoltdbFileWithNoSync(compactedDBName)
+	if is.err != nil {
+		return
+	}
+
+	for i, object := range is.sourceObjects {
+		if i == seedFileIdx {
+			continue
+		}
+		downloadAt := filepath.Join(workingDir, object.Name)
+
+		is.err = shipper_util.DownloadFileFromStorage(func() (io.ReadCloser, error) {
+			return is.baseIndexSet.GetFile(ctx, is.tableName, is.userID, object.Name)
+		}, shipper_util.IsCompressedFile(object.Name), downloadAt, false, is.logger)
+		if is.err != nil {
+			return
+		}
+
+		is.err = readFile(util_log.Logger, downloadAt, is.writeBatch)
+		if is.err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+// mustRecreateCompactedDB returns true if the compacted db should be recreated
```

Comment:
This seems unused.

Reply:
Yeah, it should be used by the index set instance as you rightly pointed out below.